File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null) return null;
        if(head.next == null ) 
          return null;

        ListNode slow = head;
        ListNode fast = head;

        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if(fast == slow) break;
        }

        if(fast == null || fast.next == null) return null;

        ListNode temp =head;
        while(temp != slow) {
            slow = slow.next;
            temp = temp.next;
        }
        return slow;
    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        // Edge case: if list is empty or has only one node → no cycle possible
        if (!head || !head->next) return nullptr;
        
        ListNode* slow = head;
        ListNode* fast = head;
        
        // Detect Cycle
        while(fast && fast->next) {
            slow = slow->next; // one step
            fast = fast->next->next; // two step

            // if slow and fast pointer meet then break the loop
            if (slow == fast) break;
        }

        // no cycle detect
        if (slow != fast) return nullptr;

        // find start of cycle
        ListNode* temp = head;
        while (temp != slow) {
            slow = slow->next;
            temp = temp->next;
        }

        return slow; // start of cycle
    }
};